File: test_c/ham_pairs.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>

/**
 * Write a program to find the least and greatest Hamming distances among all 
 * pairs of values in an array of 32 two-byte half-words. Assume all values are 
 * signed 16-bit (“half-word”) integers. The array of integers runs from data 
 * memory location 0 to 63. Even-numbered addresses are MSBs, following odd 
 * addresses are LSBs, e.g. a concatenation of addresses 0 and 1 forms a 16-bit 
 * two’s complement half-word. Write the minimum distance in location 64 and the 
 * maximum in 65.
 */

#define ARR_SIZE 6

void populate_mem(uint8_t* mem) {
    // randomly generate values to be used 
    for (int i = 0; i < ARR_SIZE; i++) {
        (*mem) = (uint8_t)((rand() & 0xFF));
        mem++;
    }
}

void to_binary(uint8_t n, char *buffer) {
    for (int i = 7; i >= 0; i--) {
        // (n >> i) & 1: Shift the bit to the 0th position and check it
        buffer[7 - i] = (n & (1 << i)) ? '1' : '0';
    }
    buffer[8] = '\0';
}

int main() {
    srand(time(NULL));
    
    uint8_t mem[ARR_SIZE + 2];
    populate_mem(mem);

    for (int i = 0; i < ARR_SIZE / 2; i++) {
        uint16_t msb = ((uint16_t)mem[2 * i]) << 8;
        char msb_bin[9];
        to_binary(mem[2 * i], msb_bin);
        uint16_t lsb = ((uint16_t)mem[2 * i + 1]);
        char lsb_bin[9];
        to_binary(mem[2 * i + 1], lsb_bin);
        int16_t val = (int16_t)(msb | lsb & 0xFF);

        printf("mem[%d] = %d (%s%s)\n", 2 * i, val, msb_bin, lsb_bin);
    }

    uint8_t min_dist = 16; // largest possible distance is 16
    uint8_t max_dist = 0;  // smallest possible distance is 0
    for (int i = 0; i < ARR_SIZE / 2; i++) {
        for (int j = i + 1; j < ARR_SIZE / 2; j++) {
            char msb1[9];
            char msb2[9];
            to_binary(mem[2*i], msb1);
            to_binary(mem[2*j], msb2);
            
            uint8_t dist = 0;
            for (int c = 0; c < 8; c++) {
                dist += (msb1[c] == '1')^(msb2[c] == '1');
            }
            char lsb1[9];
            char lsb2[9];
            to_binary(mem[2*i + 1], lsb1);
            to_binary(mem[2*j + 1], lsb2);
            
            for (int c = 0; c < 8; c++) {
                dist += (lsb1[c] == '1')^(lsb2[c] == '1');
            }

            min_dist = min_dist < dist ? min_dist : dist;
            max_dist = max_dist > dist ? max_dist : dist;
        }
    }
    mem[ARR_SIZE] = min_dist;
    mem[ARR_SIZE + 1] = max_dist;

    printf("The minimum Hamming distance is %d\n", mem[ARR_SIZE]);
    printf("The maximum Hamming distance is %d\n", mem[ARR_SIZE + 1]);

    return 0;
}

File: python_implementations/ham_pairs.py
"""
Write a program to find the least and greatest Hamming distances among all
pairs of values in an array of 32 two-byte half-words. Assume all values are
signed 16-bit ("half-word") integers. The array of integers runs from data
memory location 0 to 63. Even-numbered addresses are MSBs, following odd
addresses are LSBs, e.g. a concatenation of addresses 0 and 1 forms a 16-bit
two's complement half-word. Write the minimum distance in location 64 and the
maximum in 65.

Register allocation:
    r0  = i              (outer loop index)
    r1  = j              (inner loop index)
    r2  = min_dist
    r3  = max_dist
    r4  = curr_dist      (this pair)
    r5  = addr_msb1      (2*i)
    r6  = addr_lsb1      (2*i + 1)
    r7  = addr_msb2      (2*j)
    r8  = addr_lsb2      (2*j + 1)
    r9  = xor_msb        (MSB1 XOR MSB2)
    r10 = xor_lsb        (LSB1 XOR LSB2)
    r11 = shift_save     (temp during bit counting)
"""

import sys

from bitstring import BitArray
from util import Machine


def main(vm: Machine | None = None) -> int:
    if vm is None:
        vm = Machine()

    # r2 <- min_dist (init to 16, the maximum possible Hamming distance)
    vm.ldi_I(BitArray(uint=16, length=8))
    vm.sto_R("r2")
    # r3 <- max_dist (init to 0)
    vm.ldi_I(BitArray(uint=0, length=8))
    vm.sto_R("r3")

    for i in range(32):
        for j in range(i + 1, 32):
            vm.ldi_I(BitArray(uint=i, length=8))
            vm.sto_R("r0")
            vm.ldi_I(BitArray(uint=j, length=8))
            vm.sto_R("r1")

            # r4 <- curr_dist = 0
            vm.ldi_I(BitArray(uint=0, length=8))
            vm.sto_R("r4")

            # r5 <- addr_msb1 = 2*i;  r6 <- addr_lsb1 = 2*i + 1
            vm.mov_R("r0")
            vm.shf_I(BitArray(int=1, length=8))
            vm.sto_R("r5")
            vm.addi_I(BitArray(uint=1, length=8))
            vm.sto_R("r6")

            # r7 <- addr_msb2 = 2*j;  r8 <- addr_lsb2 = 2*j + 1
            vm.mov_R("r1")
            vm.shf_I(BitArray(int=1, length=8))
            vm.sto_R("r7")
            vm.addi_I(BitArray(uint=1, length=8))
            vm.sto_R("r8")

            # r9 <- MSB1 XOR MSB2
            vm.ld_R("r5")
            vm.sto_R("r9")
            vm.ld_R("r7")
            vm.xor_R("r9")
            vm.sto_R("r9")

            # r10 <- LSB1 XOR LSB2
            vm.ld_R("r6")
            vm.sto_R("r10")
            vm.ld_R("r8")
            vm.xor_R("r10")
            vm.sto_R("r10")

            # count set bits in r9 (MSB XOR), accumulate into r4
            vm.mov_R("r9")
            for _ in range(8):
                vm.shf_I(BitArray(int=1, length=8))
                if vm.carry_flag[0]:
                    vm.sto_R("r11")
                    vm.mov_R("r4")
                    vm.addi_I(BitArray(uint=1, length=8))
                    vm.sto_R("r4")
                    vm.mov_R("r11")

            # count set bits in r10 (LSB XOR), accumulate into r4
            vm.mov_R("r10")
            for _ in range(8):
                vm.shf_I(BitArray(int=1, length=8))
                if vm.carry_flag[0]:
                    vm.sto_R("r11")
                    vm.mov_R("r4")
                    vm.addi_I(BitArray(uint=1, length=8))
                    vm.sto_R("r4")
                    vm.mov_R("r11")

            # update min: if curr_dist < min_dist, min_dist = curr_dist
            vm.mov_R("r4")
            vm.cmp_R("r2")
            if vm.carry_flag[0]:
                vm.sto_R("r2")

            # update max: if max_dist < curr_dist, max_dist = curr_dist
            vm.mov_R("r3")
            vm.cmp_R("r4")
            if vm.carry_flag[0]:
                vm.mov_R("r4")
                vm.sto_R("r3")

    # store min_dist to mem[64], max_dist to mem[65]
    vm.ldi_I(BitArray(uint=64, length=8))
    vm.sto_R("r0")
    vm.mov_R("r2")
    vm.st_R("r0")

    vm.ldi_I(BitArray(uint=65, length=8))
    vm.sto_R("r0")
    vm.mov_R("r3")
    vm.st_R("r0")

    return 0


def test() -> None:
    import random as _random

    def _hamming(a: int, b: int) -> int:
        return bin((a ^ b) & 0xFFFF).count("1")

    def _run(values: list[int], label: str) -> None:
        vm = Machine()
        for idx, v in enumerate(values):
            vm.mem[idx * 2] = BitArray(uint=(v >> 8) & 0xFF, length=8)
            vm.mem[idx * 2 + 1] = BitArray(uint=v & 0xFF, length=8)
        main(vm)
        dists = [
            _hamming(values[i], values[j]) for i in range(32) for j in range(i + 1, 32)
        ]
        assert vm.mem[64].uint == min(dists), (
            f"{label}: min wrong (got {vm.mem[64].uint}, expected {min(dists)})"
        )
        assert vm.mem[65].uint == max(dists), (
            f"{label}: max wrong (got {vm.mem[65].uint}, expected {max(dists)})"
        )

    _run([0x0000] * 32, "all zeros")
    _run([0xFFFF] * 32, "all max")
    _run([0x0000] * 31 + [0xFFFF], "one differs maximally")
    _run([0x0000, 0xFFFF] + [0xAAAA] * 30, "two extremes plus alternating")
    _random.seed(42)
    _run([_random.randint(0, 0xFFFF) for _ in range(32)], "random seed 42")
    print("All tests passed!")


if __name__ == "__main__":
    test()
    sys.exit(main())
